### PR TITLE
Implement indexable api for topology

### PR DESCRIPTION
**Pull request by juliohm**
As discussed on Zulip

### CODE REVIEW

File: src/topologies.jl [hunk 1]
```diff
@@ -132,6 +132,26 @@ Return the number of facets of the `topology`.
...
+
+Base.lastindex(t::Topology) = nelements(t)
+
+Base.length(t::Topology) = nelements(t)
+
+Base.iterate(t::Topology, state=1) = state > nelements(t) ? nothing : (t[state], state + 1)
+
+Base.eltype(t::Topology) = eltype([t[i] for i in 1:nelements(t)])
+
+Base.keys(t::Topology) = 1:nelements(t)
```
Comment (JoshuaLampert) [inline src/topologies.jl]:
```suggestion
Base.keys(t::Topology) = Base.OneTo(nelements(t))
```
I think, often these kind of index ranges are a `Base.OneTo`, which behaves very similar to `1:`, but can sometimes be slightly more efficient?

File: src/topologies.jl [hunk 2]
```diff
@@ -132,6 +132,26 @@ Return the number of facets of the `topology`.
 """
 function nfacets(::Topology) end
 
+# ----------
+# FALLBACKS
+# ----------
+
+Base.getindex(t::Topology, ind::Int) = element(t, ind)
+
+Base.getindex(t::Topology, inds::AbstractVector) = [element(t, ind) for ind in inds]
```
Comment (JoshuaLampert) [inline src/topologies.jl]:
```suggestion
Base.getindex(t::Topology, inds::AbstractVector) = [t[ind] for ind in inds]
```

**Review by JoshuaLampert — COMMENTED** [review]
Looks good to me. I have only two minor suggestions.

File: src/topologies.jl (hunk 1 above)
Comment (juliohm) [reply src/topologies.jl]: Interesting. Didn't know there was a difference in performance.

**Review by JoshuaLampert — APPROVED** [review]

**Comment by juliohm** [comment]
I've also implemented the suggested optimizations for Domain for consistency. Waiting for tests before merging and releasing a patch.

**Comment by JoshuaLampert** [comment]
Thanks!

**Pull request merged**